server/shared/storage/amazon.ts

```diff
@@ -0,0 +1,123 @@
+import { Config } from '../../config';
+import miss from 'mississippi';
+import path from 'path';
+import S3 from 'aws-sdk/clients/s3';
+
+const isNotFound = (err: AWSError): boolean => err.code === 'NoSuchKey';
+
+class Amazon implements IStorage {
+  #bucket: string;
+  #client: S3;
+
+  constructor(config: Config) {
+    const amazonConfig = config.storage.amazon;
+
+    const s3Config = {
+      accessKeyId: amazonConfig.key,
+      secretAccessKey: amazonConfig.secret,
+      region: amazonConfig.region,
+      signatureVersion: 'v4',
+      apiVersion: '2006-03-01',
+      maxRetries: 3
+    };
+
+    this.#bucket = amazonConfig.bucket;
+    this.#client = new S3(s3Config);
+  }
+
+  // API docs: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#getObject-property
+  getFile(key: string): Promise<ContentResponse<string>> {
+    const params = { Bucket: this.#bucket, Key: key };
+    return this.#client.getObject(params).promise()
+      .then(({ Body: data }) => ({ content: data.toString(), raw: data as Buffer }))
+      .catch(err => isNotFound(err) ? null : Promise.reject(err));
+  }
+
+  // API docs: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#getObject-property
+  createReadStream(key: string): NodeJS.ReadableStream {
+    const params = { Bucket: this.#bucket, Key: key };
+    return this.#client.getObject(params).createReadStream();
+  }
+
+  // API docs: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#upload-property
+  createWriteStream(key: string): NodeJS.WritableStream {
+    const throughStream = miss.through();
+    const params = { Bucket: this.#bucket, Key: key, Body: throughStream };
+    this.#client.upload(params, () => null);
+    return throughStream;
+  }
+
+  // API docs: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#putObject-property
+  async saveFile(key: string, data: Buffer): Promise<void> {
+    const params = { Bucket: this.#bucket, Key: key, Body: data };
+    await this.#client.putObject(params).promise();
+  }
+
+  // API docs: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#copyObject-property
+  async copyFile(key: string, newKey: string): Promise<void> {
+    const { base, ...rest } = path.parse(key);
+    const encodedSource = path.format({ base: encodeURIComponent(base), ...rest });
+    const params = {
+      Bucket: this.#bucket,
+      CopySource: this.path(`/${encodedSource}`),
+      Key: newKey
+    };
+    await this.#client.copyObject(params).promise();
+  }
+
+  async moveFile(key: string, newKey: string): Promise<void> {
+    await this.copyFile(key, newKey);
+    await this.deleteFile(key);
+  }
+
+  // API docs: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#deleteObject-property
+  async deleteFile(key: string): Promise<DeleteResponse> {
+    const params = { Bucket: this.#bucket, Key: key };
+    await this.#client.deleteObject(params).promise();
+    return { isDeleted: true };
```

Comment:
Do we also need to cast provider's errors to ours so that errors are consistent between providers? It is a little bit of work so I am not sure about this one, @MiroDojkic what do you think?

Reply:
@ikovac makes sense to me, but we can leave that for improvement PR since we didn't have that interface previously.